source/opt/value_number_table.cpp

```diff
@@ -48,24 +68,111 @@ uint32_t ValueNumberTable::GetValueNumber(spvtools::ir::Instruction* inst) {
+          value = 0;
+          break;
+        }
+      }
+      if (value != 0) {
+        id_to_value_[inst->result_id()] = value;
+        return value;
+      }
+    }
   }
 
+  // Replace all of the operands by their value number.  The sign bit will be
+  // set to distinguish between an id and a value number.
+  ir::Instruction value_ins(context(), inst->opcode(), inst->type_id(),
+                            inst->result_id(), {});
+  for (uint32_t o = 0; o < inst->NumInOperands(); ++o) {
+    const ir::Operand& op = inst->GetInOperand(o);
+    if (spvIsIdType(op.type)) {
+      uint32_t id_value = op.words[0];
+      auto use_id_to_val = id_to_value_.find(id_value);
+      if (use_id_to_val != id_to_value_.end()) {
+        id_value = (1 << 31) | use_id_to_val->second;
+      }
+      value_ins.AddOperand(ir::Operand(op.type, {id_value}));
+    } else {
+      value_ins.AddOperand(ir::Operand(op.type, op.words));
+    }
+  }
+
+  // TODO: Implement a normal form for opcodes that commute like integer
+  // addition.  This will let us know that a+b is the same value as b+a.
+
   // Otherwise, we check if this value has been computed before.
-  auto value = instruction_to_value_.find(*inst);
-  if (value != instruction_to_value_.end()) {
-    uint32_t value_number = id_to_value_[value->first.result_id()];
-    id_to_value_[inst->result_id()] = value_number;
-    return value_number;
+  auto value_iterator = instruction_to_value_.find(value_ins);
+  if (value_iterator != instruction_to_value_.end()) {
+    value = id_to_value_[value_iterator->first.result_id()];
+    id_to_value_[inst->result_id()] = value;
+    return value;
   }
 
   // If not, assign it a new value number.
-  uint32_t value_number = TakeNextValueNumber();
-  id_to_value_[inst->result_id()] = value_number;
-  instruction_to_value_[*inst] = value_number;
-  return value_number;
+  value = TakeNextValueNumber();
+  id_to_value_[inst->result_id()] = value;
+  instruction_to_value_[value_ins] = value;
+  return value;
+}
+
+void ValueNumberTable::BuildDominatorTreeValueNumberTable() {
+  // First value number the headers.
+  for (auto& inst : context()->annotations()) {
+    if (inst.result_id() != 0) {
+      AssignValueNumber(&inst);
+    }
+  }
+
+  for (auto& inst : context()->capabilities()) {
+    if (inst.result_id() != 0) {
+      AssignValueNumber(&inst);
+    }
+  }
+
+  for (auto& inst : context()->types_values()) {
+    if (inst.result_id() != 0) {
+      AssignValueNumber(&inst);
+    }
+  }
+
+  for (auto& inst : context()->module()->ext_inst_imports()) {
+    if (inst.result_id() != 0) {
+      AssignValueNumber(&inst);
+    }
+  }
+
+  for (ir::Function& func : *context()->module()) {
+    // For best results we want to traverse the code in reverse post order.
+    // This happens naturally because of the forward referencing rules.
+    for (ir::BasicBlock& block : func) {
+      for (ir::Instruction& inst : block) {
```

Comment:
But, this won't guarantee dom order visits, right?  You'd need to compute structured ordering before doing this.

Reply:
Yes it does. In section 2.16.1 under the heading for the CFG is says:

The order of blocks in a function must satisfy the rule that blocks appear before all blocks they dominate.

If is have a diamond there is not guarantee that it will visit the side nodes before the merge node.  However, I am not too concerned about that case.  If we do, good, if we don't, not a problem.